casual/casual-caller/src/main/java/se/laz/casual/connection/caller/TpCallerImpl.java

```diff
@@ -0,0 +1,127 @@
+import se.laz.casual.connection.caller.pool.PoolMatcher;
+import se.laz.casual.jca.CasualConnection;
+import se.laz.casual.jca.CasualRequestInfo;
+import se.laz.casual.network.connection.CasualConnectionException;
+
+import javax.inject.Inject;
+import javax.resource.ResourceException;
+import javax.resource.spi.ConnectionRequestInfo;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+public class TpCallerImpl implements TpCaller
+{
+    private static final Logger LOG = Logger.getLogger(TpCallerImpl.class.getName());
+    private Cache cache;
+    private PoolMatcher poolMatcher;
+    private PoolManager poolManager;
+
+    // for wls
+    public TpCallerImpl()
+    {}
+
+    @Inject
+    public TpCallerImpl(PoolMatcher poolMatcher, Cache cache, PoolManager poolManager)
+    {
+        this.cache = cache;
+        this.poolMatcher = poolMatcher;
+        this.poolManager = poolManager;
+    }
+
+    @Override
+    public ServiceReturn<CasualBuffer> tpcall(String serviceName, CasualBuffer data, Flag<AtmiFlags> flags)
+    {
+        LOG.finest(() -> "tpcall<" + serviceName + ">");
+        return doCall(serviceName, connection -> connection.tpcall(serviceName, data, flags), this::tpenoentReply);
+    }
+
+    @Override
+    public CompletableFuture<ServiceReturn<CasualBuffer>> tpacall(String serviceName, CasualBuffer data, Flag<AtmiFlags> flags)
+    {
+        LOG.finest(() -> "tpacall<" + serviceName + ">");
+        return doCall(serviceName, connection -> connection.tpacall(serviceName, data, flags), () -> CompletableFuture.supplyAsync(this::tpenoentReply));
+    }
+
+    @Override
+    public boolean serviceExist(ServiceInfo serviceInfo)
+    {
+        return !poolMatcher.match(serviceInfo, poolManager.getPools()).isEmpty();
+    }
+
+    private <R> R doCall(String serviceName,  Function<CasualConnection, R> callFunction, Supplier<R> tpenoentSupplier)
+    {
+        ServiceInfo serviceInfo = ServiceInfo.of(serviceName);
+        List<CacheEntry> cacheEntries = cache.get(serviceInfo);
+        if(cacheEntries.isEmpty())
+        {
+            List<MatchingEntry> matchingEntries = poolMatcher.match(serviceInfo, poolManager.getPools());
+            cache.store(matchingEntries);
+            cacheEntries = cache.get(serviceInfo);
+            if(cacheEntries.isEmpty())
+            {
+                return tpenoentSupplier.get();
+            }
+        }
+        return makeServiceCall(cacheEntries, serviceName, callFunction);
+    }
+
+    private <R> R makeServiceCall(List<CacheEntry> matchingEntries, String serviceName, Function<CasualConnection, R> function)
+    {
+        List<CacheEntry> entries = matchingEntries.stream().collect(Collectors.toList());
+        Exception thrownException = null;
+        Collections.shuffle(entries);
```

Comment:
Don't we lose the order by hops that was established earlier in the code? Unless I'm missing something this should only shuffle parts of the list with the same hops at once, not all of the list in one go

Reply:
The list only contains entries with the lowest hops